File: src/com/timepath/quakec/ide/QCFoldingBuilder.java
package com.timepath.quakec.ide;

import com.intellij.lang.ASTNode;
import com.intellij.lang.folding.FoldingBuilder;
import com.intellij.lang.folding.FoldingDescriptor;
import com.intellij.openapi.editor.Document;
import com.intellij.openapi.util.TextRange;
import com.intellij.psi.tree.IElementType;
import com.timepath.quakec.psi.QCTypes;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.util.ArrayList;
import java.util.List;

/**
 * @author TimePath
 */
public class QCFoldingBuilder implements FoldingBuilder {
    @NotNull
    @Override
    public FoldingDescriptor[] buildFoldRegions(@NotNull ASTNode node, @NotNull Document document) {
        List<FoldingDescriptor> list = new ArrayList<FoldingDescriptor>();
        buildFolding(node, list);
        FoldingDescriptor[] descriptors = new FoldingDescriptor[list.size()];
        return list.toArray(descriptors);
    }

    private static void buildFolding(ASTNode node, List<FoldingDescriptor> list) {
        final IElementType elementType = node.getElementType();
        if (elementType == QCTypes.BLOCK || elementType == QCTypes.BLOCK_SWITCH) {
            final TextRange range = node.getTextRange();
            list.add(new FoldingDescriptor(node, range));
        }
        for (ASTNode child : node.getChildren(null)) {
            buildFolding(child, list);
        }
    }

    @Nullable
    @Override
    public String getPlaceholderText(@NotNull ASTNode node) {
        return "{...}";
    }

    @Override
    public boolean isCollapsedByDefault(@NotNull ASTNode node) {
        return false;
    }
}

File: src/main/kotlin/com/timepath/quakec/ide/QCFoldingBuilder.kt
package com.timepath.quakec.ide

import com.intellij.lang.ASTNode
import com.intellij.lang.folding.FoldingBuilder
import com.intellij.lang.folding.FoldingDescriptor
import com.intellij.openapi.editor.Document
import com.timepath.quakec.psi.QCTypes

import java.util.ArrayList

class QCFoldingBuilder : FoldingBuilder {
    override fun buildFoldRegions(node: ASTNode, document: Document): Array<FoldingDescriptor> {
        val list = ArrayList<FoldingDescriptor>()
        buildFolding(node, list)
        return list.toTypedArray()
    }

    private fun buildFolding(node: ASTNode, list: MutableList<FoldingDescriptor>) {
        val elementType = node.elementType
        if (elementType == QCTypes.BLOCK || elementType == QCTypes.BLOCK_SWITCH) {
            val range = node.textRange
            list.add(FoldingDescriptor(node, range))
        }
        for (child in node.getChildren(null)) {
            buildFolding(child, list)
        }
    }

    override fun getPlaceholderText(node: ASTNode) = "{...}"

    override fun isCollapsedByDefault(node: ASTNode) = false
}
